### ISSUE TITLE
Declaring modules in WebIDL

### ISSUE THREAD

**Issue by littledan** [heycam/webidl#592]
There's talk in various places about adding built-in modules to JavaScript and the Web Platform. For this, it could be useful to add support to WebIDL for defining a module.

`namespace`/`[LegacyNamespace=...]` could be a good model here, but lots of details are a bit different, as modules don't have all the flexibility of objects. For example, attributes in modules (which will always be readonly) shouldn't trigger any kind of evaluation/execution when they are accessed.

We could start with modules exporting only functions and interfaces (classes). It could look something like this:

```
module "std::foo" {
  double sum(double x, double y);
}
[Module="std::foo"]
interface Bar { }
```

Thoughts?

(this issue should be tagged jsidl)

**Comment by annevk** [comment]
It seems more natural to nest the interface inside the module and support "partial module" for splitting things up, if needed.

**Comment by littledan** [comment]
Well, something like that was my original intuition for namespaces as well, but we ended up going with an extended attribute in https://github.com/heycam/webidl/pull/425 ; I was trying to copy that convention. I'd be fine with this instead, though it may be more work in parsers to make it work:

```
module "std::foo" {
  double sum(double x, double y);
  interface Bar { }
}
```

Should namespaces use this pattern too, for `[LegacyNamespace]`? (I guess I'm not sure where to say "legacy", then...)

Partial modules sound important too, I agree.

**Comment by domenic** [comment]
I think extended attributes is much more natural. Web IDL doesn't have any nesting so far I believe, and that's quite nice; you just read the top-level to find all constructs. I don't see any benefits of a nesting syntax.

**Comment by bzbarsky** [comment]
> Should namespaces use this pattern too, for [LegacyNamespace]?

There are namespaces, which have dedicated syntax, and there is the `[LegacyNamespace]` thing, which is meant for some existing stuff that doesn't actually look like Web IDL namespace, right?

**Comment by littledan** [comment]
@bzbarsky I don't understand the technical difference; it seems like the main thing is that one is legacy and the other is not. What does a namespace "actually" look like?

How should we decide between @annevk's and @domenic's proposals? I am OK with either.

**Comment by annevk** [comment]
My preference is for nesting since we mostly use extended attributes for things we didn't think of initially and bolted on later. It seems to me that if we're gonna add namepacing it should be a first-class citizen syntax-wise.

**Comment by littledan** [comment]
If interfaces in namespaces becomes more widespread (we'll at least add it for Intl if that is converted to WebIDL), should we make the analogous change for namespaces?

**Comment by domenic** [comment]
I don't think extended attributes reflect bolted-on versus not. I think they reflect "modifies something" vs. not.

IMO having classes (and other constructs) be the top-level has a number of nice properties, e.g. implementers often put each class in its own file. I don't think the aesthetic benefits of nesting (which are debatable anyway; not everyone enjoys an extra level of indentation) outweigh the complexity of introducing two levels to the IDL system.

**Comment by littledan** [comment]
@domenic Would "partial module" be enough to allow file splitting?

**Comment by annevk** [comment]
I guess another way of looking at this is what the scope of the name is. Can interface names overlap due to modules and namespaces?

**Comment by littledan** [comment]
@domenic I think this relates to the question of whether an extended attribute makes sense for putting an interface in a module: if they have the same name, then don't they sort of clash, if they are both at the top level? If you have a partial interface declaration later, how do you determine which one it resolves to?

**Comment by domenic** [comment]
I can see how they might "sort of" clash on an intuitive level, but I think it's pretty easy to make a rule that partial declarations would have to have matching [Module] or [LegacyNamespace] declarations.

**Comment by bzbarsky** [comment]
Just to chime in on a pure implementation level, right now webidl has various scopes that constrain name resolution (toplevel, dictionary, interface) and all those scopes map to reasonable syntactic constructs.  Introducing scopes that don't map to a simple syntactic construct would lead to some amount of parser implementation complexity which may not be warranted...

**Comment by littledan** [comment]
@bzbarsky Do you mean, in practice, it would be easier to distinguish the scope using an extended attribute?

**Comment by bzbarsky** [comment]
No, in practice (at least in the Gecko parser) it would be easier to distinguish the scope by having an explicit syntactic construct with nesting for that scope, like `namespace foo { ... }` or whatnot.

**Comment by domenic** [comment]
That's helpful feedback to have. How does that weigh against the issues around the traditional one interface per file, as discussed in https://github.com/heycam/webidl/issues/592#issuecomment-444523542 ?

**Comment by littledan** [comment]
Some notes on how it might make sense to specify modules in WebIDL:
- Modules can be created in JavaScript with this mechanism that @domenic proposed: https://github.com/tc39/proposal-javascript-standard-library/pull/44
- I think it'd make sense to have WebIDL initialize the module map, given that import-maps' current draft specification is based on modifying the "resolve a module specifier" algorithm, see https://wicg.github.io/import-maps/#resolving .
- We may need a separate entry-point into HTML; right now, we rely on being called when a new global object is created, but we'd also want to be called separately when a new module map is created (and given the realm as a parameter).

Any concerns with this outline?

**Comment by domenic** [comment]
It seems like a goal of the design is to allow reuse of names between globals and modules. But, I don't think the current draft PR has this quite figured out. In particular, consider the situation

```webidl
module "std:dom2" {
  interface Node { };

  interface Foo {
    void op(Node n);
  };
 
  interface Bar : Node {};

  interface mixin Baz;
  Node includes Baz;
};

module "std:dom3" {
  interface Node {};
};
```

In these situations, which `Node` is being referred to in the various usage sites?

It seems like there's a lot of assumptions currently baked in to Web IDL about there being one global name table :-/. Probably we'd start needing notions of "fully qualfied name" (like `"std:dom2"/Node` or something).

**Comment by littledan** [comment]
Yeah, this is definitely an area of the PR that needs work; how partial interfaces in partial modules actually accumulate the right members (and not members outside the module) is another. I'm thinking to define both partials and name resolution imperatively as building blocks in separate PRs to make this possible.

As for a notion of fully qualified name, how would [qualified name](https://heycam.github.io/webidl/#qualified-name) be? This is what determines the name for `@@toStringTag`, but it seems like it'd be unique enough to me.

Do you think there are ever use cases where you'd shadow a name defined globally with one that's local to a module, and still want to refer to the global one within that module? If not, we can save ourselves a bit of hassle by not needing to define IDL syntax for these cases.

**Comment by domenic** [comment]
> As for a notion of fully qualified name, how would qualified name be? This is what determines the name for @@toStringTag, but it seems like it'd be unique enough to me.

Interesting. I see per the PR the @@toStringTag for the examples would be `"std:dom2.Node"`?I'm not sure how I feel about that, although in general I'm not sure how I feel about `@@toStringTag` usage at all. (I vaguely would prefer we stopped using it for new things, maybe.)

> Do you think there are ever use cases where you'd shadow a name defined globally with one that's local to a module, and still want to refer to the global one within that module?

I can definitely think of cases, e.g. defining a modern type in a module and wanting to add a utility method to convert from the legacy global type to the modern type. But, I think it is fine for an MVP to not allow such references, and add syntax for them when we actually encounter such cases.
